Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,206 @@
+import Foundation
+
+/// Corpus & Scheduler based on 
+/// Coverage-based Greybox Fuzzing as Markov Chain paper
+/// https://mboehme.github.io/paper/TSE18.pdf
+/// Simply put, the corpus keeps track of which paths have been found, and prioritizes seeds
+/// whose path has been hit less than average. Ideally, this allows the fuzzer to prioritize
+/// less explored coverage.
+
+public class MarkovCorpus: ComponentBase, Corpus {
+
+    // All programs currently in the corpus
+    private var allIncludedPrograms: [Program] = []
+    // Queue of programs to be executed next, all of which hit a rare edge
+    private var programExecutionQueue: [Program] = []
+
+    // For each edge encountered thus far, track which program initially discovered it
+    private var edgeMap: [UInt32:Program] = [:]
+
+    // This scheduler tracks the total number of samples it has returned
+    // This allows it to build an initial baseline by randomly selecting a program to mutate
+    // before switching to the more computationally expensive selection of programs that
+    // hit infreqent edges
+    private var totalExecs: UInt32 = 0
+
+    // This scheduler returns one base program multiple times, in order to compensate the overhead caused by tracking
+    // edge counts
+    private var currentProg: Program
+    private var remainingEnergy: UInt32 = 0
+
+    // Markov corpus requires an evaluator that tracks edge coverage
+    // Thus, the corpus object keeps a reference to the evaluator, in order to only downcast once
+    private var covEvaluator: ProgramCoverageEvaluator
+
+    // Rate at which selected samples will be included, to promote diversity between instances
+    // Equivalent to 1 - dropoutRate
+    private var dropoutRate: Double
+
+    // The scheduler will initially selectd the 1 / desiredSelectionProportion samples with the least frequent
+    // edge hits in each round, before dropout is applied
+    private let desiredSelectionProportion = 8
+
+    public init(covEvaluator: ProgramCoverageEvaluator, dropoutRate: Double) {
+        self.dropoutRate = dropoutRate
+        covEvaluator.enableEdgeTracking()
+        self.covEvaluator = covEvaluator
+        self.currentProg = Program()
+        super.init(name: "MarkovCorpus")
+    }
+
+    override func initialize() {
+        assert(covEvaluator === fuzzer.evaluator as! ProgramCoverageEvaluator)
+        currentProg = makeSeedProgram()
+        add(currentProg)
```

Comment:
And then how about we actually execute and evaluate the seed program so proper aspects are available for it? I guess that could happen either in a shared helper function in `Corpus` or it could be taken care of in `Fuzzer`?

Reply:
I've added the following short segment, using on a helper function here: https://github.com/googleprojectzero/fuzzilli/pull/171/files/79c0896fdcc6d49da4dbab8af446137af6c3e3af#diff-77dad126ed05a8ccbc83ec9621d3fca2faff468dcc392724709fda4ad0ad0687R54


``` swift
let aspects = execAndEvalProg(currentProg)! 
add(currentProg, aspects)
```